tests/testthat/_snaps/constructor.md

```diff
@@ -69,3 +69,23 @@
       new_object(foo(...), y = y)
       <environment: 0x0>
 
+# can create constructors with missing or lazy defaults
+
+    Code
+      Person()
+    Condition
+      Error:
+      ! <Person> object properties are invalid:
+      - @first_name must be <character>, not <NULL>
+      - @last_name must be <MISSING> or <character>, not <NULL>
+      - @nick_name must be <character>, not <NULL>
+
+---
+
+    Code
+      Person("Alice")
+    Condition
+      Error:
+      ! <Person> object properties are invalid:
+      - @last_name must be <MISSING> or <character>, not <NULL>
```

Comment:
This seems a bit confusing because it is missing?

Reply:
The issue happens because `person@last_name <- <value>` is never set since `<value>` is missing. Then, when the validator fetches `person@last_name`, it returns `NULL` (because `attr(x, "does_not_exist")` return `NULL`), which is the wrong type and causes the error.

This whole test probably needs to be rewritten to make sense with the new behavior.

Reply:
https://github.com/RConsortium/S7/blob/15a01a3ab76848981ca03ab30e30e3609c554a01/src/prop.c#L260-L261

Reply:
If it's not set, should we be using `quote(expr = )`?

Reply:
To do that, we would have to use `Rf_setAttrib(object, Rf_install("name"), R_missingArg)` in `prop<-` when the object is first constructed.

`quote(expr=)` is such a pain to work with at the R level; I'm beginning to think that perhaps we should not create a pattern that requires all object instances to be handled with care, checking with `missing()` like this.

To make the "deprecate via getter+setter" pattern possible, maybe we should return to what we discussed in https://github.com/RConsortium/S7/pull/396#discussion_r1486563418 and altogether bypass or not invoke custom setters on initial construction.

We would then need to provide a convenient way to "opt-in" to running the setters without requiring a full custom constructor. 

Perhaps we add an argument to `new_class(..., initialize = function(self) {})`, a function authors can provide. The `initialize` function will be called with the output of the `constructor` after the `constructor()` has returned but before `validator()` is called. (somewhat analogous to `__new__` and `__init__` in Python.)

Reply:
We may also want to add an `initializer` arg to `new_property`, defaulting to `setter`.
```r
new_property(..., initializer = setter)
```

Reply:
Could we just use `class_missing` as the sentinel value? Or perhaps something similar but named like `arg_missing`?

Reply:
I think we got a bit side-tracked into a much bigger topic than I was thinking. Looking at this with fresh eyes, I think all that I want is for the error message to be:

```
! <Person> object properties are invalid:
- @last_name must be <character>, not <NULL>
```